Housekeeping Ability for Rasenbot #4

stephenlee91 opened the pull request:
Created the ability to require new users to read and accept some rules before joining the server based on clicking a thumbs up emoji of the greetings message. This ability allows anyone who clicks on the emoji to receive a role that is hardcoded into greetings.py that allows them to access the server.

Removed autoassign.py since it is no longer needed because of the new housekeeping ability.

Changed some comments to allow for better readability.

IMPORTANT:
Permissions must be set a certain way. (Might be best to add this to documentation for the bot)
DEFAULT PERMISSIONS for @everyone:
- everything must be turned off

DEFAULT PERMISSIONS for the role that is required to access server:
- able to view messages and send messages and connect to VC, etc.

PERMISSIONS FOR TEXT CHANNEL THAT SENDS GREETINGS MESSAGE for @everyone:
- able to view messages
- able to view message history
- everything else turned off

PERMISSIONS FOR TEXT CHANNEL THAT SENDS GREETINGS MESSAGE for role to access server:
- everything turned off
(this is so that they won't ever have to see the channel again and get spammed)

PERMISSIONS FOR ALL OTHER TEXT CHANNELS AND VC CHANNELS for @everyone:
- everything must be turned off

PERMISSIONS FOR ALL OTHER TEXT CHANNELS AND VC CHANNELS for role to access server:
- able to read and view message history
- able to connect to voices

**This ability may get updated in the future if we want to change some of the rules or create a better way so that a new user does not have to see previous welcome messages and only sees one greetings message or if we want to make the message more visually appealing.**

stephenlee91 committed:
Removed auto assign and changed greetings.py to create a greetings message that requires new users to react to message to receive a role to access the server. Also changed formatting of some comments to make more readable

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -1,27 +1,40 @@
     async def on_member_join(self, member):
-        #Sends the embedded message to the channel that is the welcome page
-        channel = member.guild.system_channel
+        channel = self.client.get_channel(1011049173442895902)  # Replace with the actual ID of your welcome text channel
```
hardcoding channel ID is probably not a good idea. we should make a configuration file that holds this value.

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -1,27 +1,40 @@
+            self._last_member = member
             embed.set_author(name="RasenBot", icon_url="https://avatars.githubusercontent.com/u/39735649?s=280&v=4")
-            #A thumbnail that is displayed in the right side of the embedded message
             embed.set_thumbnail(url="http://www.willpap-projects.com/Docus/Educational/Slide_Shows/Gifs/Funny/Cat_Paws_79.gif")
```
might be good to download/include this gif in the project (ensure that it's not copyrighted) and then we don't have to ping a website. also could be a point of failure if this url dies somehow.

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -1,27 +1,40 @@
+                guild = self.client.get_guild(payload.guild_id)
+                member = guild.get_member(payload.user_id)
+                if member is not None and not member.bot:  # Check if the member exists and is not a bot
+                    role = guild.get_role(1120501110407434301) # Replace with the actual ID of the role you want to assign
```
ditto

stephenlee91 committed:
Added new suggestions: Constant greetings message that requires new users to react to. Created constantvariables.py that contains variables that pertains to each server. Created a media folder full of copyright free images and gifs for use. Created the ability to unreact to welcome message and remove role

stephenlee91 commented:
**### NEW EDIT TO CODE**

Created the ability to require new users to read and accept some rules before joining the server based on clicking a thumbs up emoji of the greetings message. This ability allows anyone who clicks on the emoji to receive a role that is hardcoded into constantvariables.py that allows them to access the server. This greetings message can be constant and is sent by typing the command, '$welcome'. If a user unreacts to the message they lose the role.

Removed autoassign.py since it is no longer needed because of the new housekeeping ability.

Changed some comments to allow for better readability.

Added a media folder that contains non-copyright images and gifs

Added a constantvariables.py file that contains variables that are each unique to a server

IMPORTANT:
Permissions must be set a certain way. (Might be best to add this to documentation for the bot)
DEFAULT PERMISSIONS for @everyone:
- everything must be turned off

DEFAULT PERMISSIONS for the role that is required to access server:
- able to view messages and send messages and connect to VC, etc.

PERMISSIONS FOR TEXT CHANNEL THAT SENDS GREETINGS MESSAGE for @everyone:
- able to view messages
- able to view message history
- everything else turned off

PERMISSIONS FOR TEXT CHANNEL THAT SENDS GREETINGS MESSAGE for role to access server:
- everything turned off
(this is so that they won't ever have to see the channel again and get spammed)

PERMISSIONS FOR ALL OTHER TEXT CHANNELS AND VC CHANNELS for @everyone:
- everything must be turned off

PERMISSIONS FOR ALL OTHER TEXT CHANNELS AND VC CHANNELS for role to access server:
- able to read and view message history
- able to connect to voices

This ability may get updated in the future if we want to change some of the rules or if we want to make the message more visually appealing.

LinsaneFC commented:
Good information for the permission stuff set in discord. Let's keep that here (ticket will be in closed after we merge) and let Benji or Anthony work on starting up the documentation. We can then move this stuff over into the wiki.

LinsaneFC commented:
Let's open up a ticket for the future changes with what can be added (just some ideas) doesn't need to be set but just so that we don't forget. You can go off what you know about it right now such as making message more visually appealing or changing rules.

TimothyYLin commented on constantvariables.py:
```diff
@@ -0,0 +1,7 @@
+Change the variables to whatever your server IDs are. Each is labelled accordingly
+'''
+
+welcome_channel_id = 1011049173442895902 # Replace with the actual ID of your welcome text channel
```
let's do all caps for constants. it's a common naming convention for constants

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,7 +1,7 @@
 from discord.ext import commands
 
-#Command that moves one user in a voice channel into another voice channel specified by the user
+# Command that moves one user in a voice channel into another voice channel specified by the user
```
Typically let's not put in nonrelated changes but for this PR it's fine especially cuz this is just a formatting thing

TimothyYLin commented on media/copyrightfreegif.gif:
most likely will have more copyrightfree gifs lol. maybe name it like discordlogo.gif or something? we'll know ourselves they're copyright free which should be enough for us

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -1,27 +1,54 @@
+            await message.add_reaction("👍")
+
+    @commands.Cog.listener()
+    async def on_raw_reaction_add(self, payload):
```
is there any way to run this without having to run it per message react?

TimothyYLin commented on constantvariables.py:
```diff
@@ -0,0 +1,7 @@
+
+welcome_channel_id = 1011049173442895902 # Replace with the actual ID of your welcome text channel
+
+verified_member_id = 1120501110407434301 # Replace with the actual ID of the role you want for the greetings message
```
ditto

LinsaneFC commented:
Some of the comments are not "under review". That was my bad for choosing them to be independent comments. Take them as if they're review comments.

stephenlee91 committed:
Created more efficiency by only running the reaction add function if the message id is the one reacted to so it reduces the number of unnecessary checks and actions

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -28,10 +28,11 @@ async def welcome(self, ctx):
     @commands.Cog.listener()
     async def on_raw_reaction_add(self, payload):
-        if payload.channel_id == constantvariables.welcome_channel_id:
+        if payload.channel_id == constantvariables.welcome_channel_id and payload.message_id == self.welcome_message_id:
```
nit: better to put that conditional before the other one since it is much less likely that the message_id will match

However, seeing a potential failure point here. If two people join back to back without the first person reacting to the message yet, since we're doing a message per person, then if the first person reacts late, this won't run as expected right? Am I missing something here?

TimothyYLin commented on cogs/greetings.py:
```diff
@@ -41,7 +42,7 @@ async def on_raw_reaction_add(self, payload):
     async def on_raw_reaction_remove(self, payload):
-        if payload.channel_id == constantvariables.welcome_channel_id:
+        if payload.channel_id == constantvariables.welcome_channel_id and payload.message_id == self.welcome_message_id:
             if "👍" in str(payload.emoji):
```
👍 seems to be a "hardcoded" emoji that can be changed/customized per server. Let's make this a constant as well.

stephenlee91 committed:
Added emoji to constantvariables, more efficiency for bot as it is able to remember welcome message after it goes offline

stephenlee91 commented:
Resolved request changes

stephenlee91 committed:
Added a readme to show purpose of folder to developers

stephenlee91 committed:
Removed READ.txt

benjaminl4917 reviewed (approved):
Good work Steven!
